Migrate composer tutorial (#3247)
Qiskit/documentation: docs/guides/composer.mdx

---
title: Explore gates and circuits with IBM Quantum Composer
description: IBM Quantum Composer is a graphical programming tool. Drag and drop operations to build circuits and run them on quantum hardware.
platform: cloud
---

{/* cspell:ignore gatename, ungroup, ungrouping */}
# Explore gates and circuits with IBM Quantum Composer

## What is it?

[IBM Quantum&reg; Composer](https://quantum.cloud.ibm.com/composer) is a graphical quantum programming tool that lets you drag and drop operations to build quantum circuits and run them on quantum hardware.

## What can it do?

**Visualize qubit states**

See how changes to your circuit affect the state of qubits, shown as an interactive q-sphere, or histograms showing measurement probabilities or statevector simulations.

**Run on quantum hardware**

Run your circuits on real quantum hardware to understand the effects of device noise.

**Automatically generate code**

Instead of writing code by hand, automatically generate OpenQASM or Python code that behaves the same way as the circuit you created with the Composer.

## Tour of the interface

> IBM Quantum Composer has a customizable set of tools that allow you to build, visualize, and run quantum circuits on quantum processing units (QPUs). Use the More options menu on each window to access additional tools and actions.

Reviewer: Should this be in quotes? I think it would make the English & translations clearer

```suggestion
IBM Quantum Composer has a customizable set of tools that allow you to build, visualize, and run quantum circuits on quantum processing units (QPUs). Use the "More options" menu on each window to access additional tools and actions.
```

Reviewer: Ditto the rest of the file, like "View"

Reviewer 2: To discuss with Becky when she gets back. I went through and put quotes around any menu/button label that uses a phrase rather than a single word ("More options", "Visualizations seed", etc.) and wasn't already emphasized otherwise in the sentence.

4. (Optional) Customize your workspace:
   * Use the **View** menu to change from the default theme to a monochrome theme. You can also select which panels to include on your workspace, then use the menu in the right corner of any panel to access options for further customization. The options to show or hide phase disks, choose the alignment of the qubits on your circuit, and reset the workspace to the default are in the View menu as well.
   * Switch between dark and light workspace themes in the bottom-right corner of the footer.

To build a circuit, you can either [drag and drop operations](#2-build-your-circuit-with-drag-and-drop), or you can enter [OpenQASM code](#3-build-your-circuit-with-openqasm-code) into the code editor.

### 2. Build your circuit with drag-and-drop

#### Operations catalog

Drag and drop operations from the operations catalog onto the quantum and classical registers. Click the search icon and enter a term in the search bar to quickly find an operation.

Collapse and expand the operations catalog by clicking the icon in the upper right corner of the operations panel. Click the icon next to it to toggle between a grid and list view of the catalog.

> Right-click an operation icon and select Info to view the definition of an operation, along with its QASM reference.

Reviewer: Does IBM want us to call it right-click or secondary-click? Fwiw, Apple calls it a secondary click in their UI, I think.

Reviewer 2: IBM Style uses "right-click".

You can now drag and drop the new operation throughout your circuit. Note that the operation is saved to this circuit but does not appear in the operations catalog for other circuits.

You can also build a custom operation directly in the OpenQASM code editor; see [Create a custom operation in OpenQASM](#create-a-custom-operation-in-openqasm) for more information.

#### Ungroup a custom or predefined operation

To ungroup the gates within a custom or predefined operation, click the operation on the Composer and select Ungroup from the contextual menu. You can now move the separate operations individually. When you ungroup an operation, each element in the former group executes independently, which may mean they execute in a different order from when they were grouped together.

#### Expand an operation’s definition

To view the operations that constitute a custom or predefined operation without ungrouping them, click **Expand definition** from the contextual menu to see the defining gates. Click the icon again to collapse the definition.

#### Rename or delete a custom operation

> To rename or delete a custom operation, right-click the operation in the operations catalog and select Rename or Delete. Deleting a custom operation from the operations catalog also deletes any instances of it on your circuit; however, deleting a custom operation from the circuit does not delete it from the operations catalog.

Reviewer: I had to reread the second sentence a few times. (Granted, I'm fatigued right now hah). I think what trips me up is the `however` making me want to have a strong connection between the two clauses. Thoughts on splitting it out so the last clause is a standalone paragraph? 

>  To rename or delete a custom operation, right-click the operation in the operations catalog and select Rename or Delete. Deleting a custom operation from the operations catalog also deletes any instances of it on your circuit.

> Note that deleting a custom operation from the circuit itself does not remove it from the operations catalog.

Maybe even make the second sentence more verbose to really spell it out, like "and you must instead..."

Read more about the phase disk visualization [here.](#phase-disk)

#### Export a circuit image

To export an image of your circuit, select File → Export circuit image. The Export options window opens, where you can choose a theme (light, dark, white on black, or black on white), a format (.svg or .png), and whether you want to apply a line wrap. After you’ve chosen your options, click Export.

### 3. Build your circuit with OpenQASM code

<Admonition type="note">
IBM Quantum Composer currently supports OpenQASM 2.0.
</Admonition>

* To open the code editor, click View → Panels → Code Editor.
* See the [Composer operations glossary](#composer-operations-glossary) for OpenQASM references to gates and other operations.
* You can define your own custom operations; see [Create a custom operation in OpenQASM](#create-a-custom-operation-in-openqasm).

> * For more information on using the OpenQASM language, including sample lines of code, see the original research paper, [Open Quantum Assembly Language](https://arxiv.org/abs/1707.03429). The [table](https://arxiv.org/pdf/1707.03429.pdf#page=9) of OpenQASM language statements from the paper is reproduced below. The OpenQASM grammar can be found in [Appendix A](https://arxiv.org/pdf/1707.03429.pdf#page=21) of the paper.

Reviewer: We don't want to link to https://quantum.cloud.ibm.com/docs/en/guides/introduction-to-qasm?

Reviewer 2: Ha, it didn't exist when this was written 🤣

</tr>
<tr>
<td><code>barrier qargs;</code></td>
<td>Prevent transformations across this source line</td>
<td><code>barrier q[0],q[1];</code></td>
</tr>
</tbody></table>

\[a] This must appear as the first non-comment line of the file.

\[b] The parameters theta, phi, and lambda are given by *parameter expressions*; for more information, see [page 5 of the paper](https://arxiv.org/abs/1707.03429#page=5) and [Appendix A](https://arxiv.org/pdf/1707.03429.pdf#page=21).

#### Create a custom operation in OpenQASM

You can define new unitary operations in the code editor (see the figure below for an example). The operations are applied using the statement `name(params) qargs;` just like the built-in operations. The parentheses are optional if there are no parameters.

To define a custom operation, enter it in the OpenQASM code editor using this format: `gatename(params) qargs;`. If you click **+Add** in the operations list, you will be prompted to enter a name for your custom operation, which you can then build in the code editor.

Once you have defined your custom operation, drag it onto the graphical editor and use the edit icon to fine-tune the parameters.

> a) The gates to be included in the custom operation:

Reviewer: I was going to suggest using a dedicated list, but I realize the images will make that tricky. Would it make more sense to use captions instead on the images? https://github.com/Qiskit/documentation/blob/main/mdx-guide.md#images

Reviewer 2: <img width="767" alt="Screenshot 2025-06-23 at 7 46 11 PM" src="https://github.com/user-attachments/assets/00652031-5b01-4a38-b0a0-3f058c26b673" />
 
Does this seem better?

Reviewer: Yes on desktop view! Maybe not as nice for mobile / small screens because there could be horizontal scroll. Why not use captions?

* To close the Inspector and return to editing your circuit, click the Inspect toggle in the toolbar.

<Admonition type="note" title="Randomness in the simulator">
The simulator creates randomness by generating results based on a *seed*. The seed is the initial value introduced into the algorithm that generates pseudorandom numbers. You can see the seed number by selecting Visualizations seed from the Edit menu. You can also set the seed yourself by changing the value in the box.
</Admonition>

## Run circuits and view results

Follow the steps below to run quantum circuits on QPUs or simulators, and view the results.

### Choose your job settings

Click **Run circuit** in the upper right corner. In the window that opens, select an available QPU. You can also choose an instance, which is associated with a plan (such as the Open, Flex, or Premium Plans). The instance you choose affects which QPUs are available to you. Click the View details link in the table of QPUs to see more information on each QPU.

Next, you can set the number of shots (executions) of your circuit that the backend will perform.

> You can optionally name the job and add tags in this panel as well. This will not change the name of the circuit. A "Composer" tag has been pre-populated to make it easier to filter your Workloads table by Composer jobs. You can remove this tag.

Reviewer: Is the tense right?

```suggestion
You can optionally name the job and add tags in this panel as well. This will not change the name of the circuit. A "Composer" tag will be pre-populated to make it easier to filter your Workloads table by Composer jobs. You can remove this tag.
```

Reviewer 2: Rewrote

![purity of the qubit state](/docs/images/guides/composer/425d4355-6a67-45ec-989c-afc2aa800448.avif)

### Probabilities view

**8-qubit limit**

This view visualizes the probabilities of the quantum state as a bar graph. The horizontal axis labels the computational basis states. The vertical axis measures the probabilities in terms of percentages. In this view the quantum phases are not represented, and is therefore an incomplete representation. However, it is useful for predicting the outcomes if each qubit is measured and the value stored in its own classical bit.

Consider the following quantum circuit and its probabilities view:

![a quantum circuit](/docs/images/guides/composer/58f83286-9707-4e43-893e-2bd161c744cd.avif)

> A circuit consisting of a column of Hadamards that creates an equal superposition of the computational basis states, followed by a two-qubit controlled-Z (CZ) gate.

Reviewer: Is this meant to be a caption? If so, it should probably move to the image itself by using the caption syntax

![measurement probabilities view](/docs/images/guides/composer/4d79349d-dadd-4e9c-8571-4617601d22f3.avif)

> A quantum circuit and its probabilities view.

Reviewer: Ditto using image caption syntax

The nodes are laid out on the q-sphere so that the basis state with all zeros (e.g., $|000\rangle)$ is at its north pole, and the basis state with all ones (e.g., $|111\rangle$) is at its south pole. Basis states with the same number of zeros (or ones) lie on a shared latitude of the q-sphere (e.g., $|001\rangle, |010\rangle, |100\rangle$). Beginning at the north pole of the q-sphere and progressing southward, each successive latitude has basis states with a greater number of ones; the latitude of a basis state is determined by its [Hamming distance](https://en.wikipedia.org/wiki/Hamming_distance) from the zero state. The q-sphere contains complete information about the quantum state in a compact representation.

Consider the following quantum circuit and its q-sphere:

![a quantum circuit](/docs/images/guides/composer/58f83286-9707-4e43-893e-2bd161c744cd.avif)

![q-sphere view](/docs/images/guides/composer/55d58744-c95c-44b3-ba29-3c7e24ec4dcc.avif)

> A quantum circuit and the q-sphere representing the state the circuit creates.

Reviewer: Is this a caption?

</Admonition>

### Statevector view

**6-qubit limit**

It is common to call $\sqrt{p_k}e^{i\varphi_k}$ the quantum amplitude. This view visualizes the quantum amplitudes as a bar graph. The horizontal axis labels the computational basis states. The vertical axis measures the magnitude of the amplitudes ($\sqrt{p_k}$) associated with each computational basis state. The color of each bar represents the quantum phase (${\varphi_k}$).

Consider the following quantum circuit and its statevector view:

![a quantum circuit](/docs/images/guides/composer/58f83286-9707-4e43-893e-2bd161c744cd.avif)

![statevector view](/docs/images/guides/composer/1f828b26-e332-4ef2-b8f0-649030fabab4.avif)

> The statevector at the terminus of the above circuit. The color wheel maps phase angle to color. The output state is expressed as a list of complex numbers.

Reviewer: Is this a caption?

## Composer operations glossary

This page is a reference that defines the various classical and quantum operations you can use to manipulate qubits in a quantum circuit. Quantum operations include quantum gates, such as the Hadamard gate, as well as operations that are not quantum gates, such as the measurement operation.

Each entry below provides details and the OpenQASM reference for each operation. See the topic on [Build your circuit with OpenQASM code](#3-build-your-circuit-with-openqasm-code) for more information.

The q-sphere image in each gate entry below shows the state after the gate operates on the initial equal superposition state $\frac{1}{\sqrt{2^{n}}}\sum_{i=0}^{2^{n}-1}|i\rangle$, where $n$ is the number of qubits needed to support the gate. See the [q-sphere](#q-sphere-view) topic for more information on this visualization.

You can define a custom operation to use in IBM Quantum Composer. For instructions, see the [Create a custom operation in OpenQASM](#create-a-custom-operation-in-openqasm) topic.

> <Admonition type="note">

Reviewer: Should this have a title like "Gate colors"?

| ------------------------------------------------- | ------------------ | ---------------------------------------------------------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| ![z\_gate](/docs/images/guides/composer/941c8820-65bc-444c-890b-b445b9958085.avif) | `z q[0];`          | ![image23](/docs/images/guides/composer/84b92ecb-3b62-4cb3-9c01-549866681816.avif)                              | The q-sphere representation shows the state after the gate operates on the initial equal superposition state $\frac{1}{\sqrt{2^{n}}}\sum_{i=0}^{2^{n}-1}\vert i\rangle, $ where $n$ is the number of qubits needed to support the gate. |

**$T^{\dagger}$ gate**

Also known as the Tdg or T-dagger gate.

> The inverse of the T gate.

Reviewer: These sentences confuse me a little because they are not complete sentences. Should they use `:`?

Reviewer 2: I think it's ok - they are typical fragments of a reference/glossary-type nature. But if we get more confusion about them in the future we'll rewrite.
